pkg/arvo/sys/hoon.hoon

```diff
@@ -910,31 +900,42 @@
     (sub len (met boz dat))
   (swp boz dat)
 ::
-::  Like `rip` but produces n-bit blocks instead of 2^n bit blocks.
-::
-++  ripn
-  ~/  %ripn
-  |=  [bits=@ud x=@]
-  ^-  (list @)
-  ?:  =(0 x)  ~
-  [(end 0 bits x) $(x (rsh 0 bits x))]
-::
 ++  rip                                                 ::  disassemble
   ~/  %rip
-  |=  [=bloq x=@]
+  |=  [a=bloq b=@u c=@]
   ^-  (list @)
-  ?:  =(0 x)  ~
-  [(end bloq 1 x) $(x (rsh bloq 1 x))]
+  ?:  =(0 c)  ~
+  [(end a b c) $(c (rsh a b c))]
 ::
 ++  rsh                                                 ::  right-shift
   ~/  %rsh
   |=  [a=bloq b=@u c=@]
   (div c (bex (mul (bex a) b)))
 ::
+++  run                                                 ::  +turn into atom
+  ~/  %run
+  |=  [a=bloq b=@u c=@ d=$-(@ @)]
+  ^-  @
+  (rep a b (rut a b c d))
+::
+++  rut                                                 ::  +turn into list
+  ~/  %rut
+  |=  [a=bloq b=@u c=@ d=$-(@ @)]
+  ^-  (list @)
+  (turn (rip a b c) d)
+::
+++  sew                                                 ::  stitch into
+  ~/  %sew
+  |=  [a=bloq [b=@u c=@u d=@] e=@]
+  ^-  @
+  %^  cat  a
+    (can a b^e c^d ~)
+  (rsh a (add b c) e)
```

Comment:
I think the 0 semantics you want here are
```suggestion
  %^  mix  e
    (can a b^e c^d ~)
 (end a (add b c) e)
```
or sth; in particular, `(sew 3 [2 1 %a] '\00\00\00\00b')` should definitely be `'\00\00a\00b'` not `'\00\00ab'`.

(Arguably `c` should also be limited so the operation cannot _increase_ the `met a` from e to the result, but it definitely should never decrease it)

Reply:
That's a good point, but a bad example:

```
> `@t`(sew 3 [2 1 %a] '\00\00\00\00b')
'\00\00a\00b'
> `@ux`'\00\00\00\00b'
0x62.0000.0000
> `@ux`(sew 3 [2 1 %a] '\00\00\00\00b')
0x62.0061.0000
```

A better example would be:

```
> `@ux`(sew 3 [2 1 %$] '\00\00\00\00b')
0x6200
> `@t`(sew 3 [2 1 %$] '\00\00\00\00b')
'\00b'
```

which should roundtrip. I had switched to `+cat` to avoid measuring `e` or combining right and left shifts, but that's just wrong here.

I think this gate is strictly more useful if it doesn't limit `c` to avoid growing `e` -- `+cut` (and many other operations) treat atoms as if they were of infinite length (note that `+can` already truncates `d` to `c` bloqs). The runtime can mutate the storage behind `e` if the usual conditions apply (on our road, refcount 1, &c) and the result would fit in `e`. Or it can realloc in place, or allocate and copy.

Reply:
You're right, `e` has conceptually infinite trailing zeroes(as a practical matter, it might be a 32-bit word which is getting `sew 0`n/`sew 3`d) and there's no reason to prevent writes to those.